oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,370 @@
+=============
+
+We have no clean, low-overhead way of composing components from multiple MFEs into a single page, or extending an MFE with additional functionality.
+
+The reality of MFEs is that while each application may represent a cohesive domain or bounded context, sometimes you need content and functionality from more than one domain on the page at the same time.
+
+Alternately, providers may want to show different _versions_ of MFEs to different users while keeping the rest of the app (header, navigation, other MFEs) unchanged.
+
+We aren't without options here, but what we have today has its own issues. It's worth walking through what we have today.
+
+Shared Libraries (Current Option #1)
+------------------------------------
+
+Because each MFE is siloed from each other - both in repositories and at runtime - our only viable mechanism for sharing code has historically been to extract it into a shared library that gets bundled with each at build time. This creates more repository/dependency overhead.
+
+Build-time package overrides (Current Option #2)
+------------------------------------------------
+
+NPM and package.json allow operators to override dependency resolution by installing an alternate version of a dependency prior to build time. This has historically been how we've allowed operators to override the header, footer, and brand.
+
+The system is confusing and somewhat brittle and only works at build time. If an operator needs different headers/footers/brands for different clients, we've just multiplied the number of builds they need to do (already a problem even without this - see the "Build time" issue above!)
+
+Frontend Plugins (Current Option #3)
+------------------------------------
+
+With the creation of the `frontend-plugin-framework <https://github.com/openedx/frontend-plugin-framework>`_, we're starting to have the capability to share components across MFEs as plugins, either at build time (direct plugins) or runtime (iframe plugins)
+
+Direct plugins create some flexibility, but couple our repositories' builds together. Iframes are good for sandboxing and isolating code, but they're a very inefficient way to compose a UI, especially given the consistency and UX/DX concerns raised above. In a way, they exacerbate the problem even more. There's no way to do direct plugins via the MFE config API, since they rely on importing modules directly into the build.
+
+Specification
+************
+
+Fundamentally, runtime module loading and shared dependencies address the majority of these problems, so our approach centers on enabling these two things.
+
+Both are the primary value proposition of module federation. Module federation is an advanced technique designed specifically for use cases like ours, and encompasses a few related ideas. It can be accomplished manually by leveraging Webpack externals, via Webpack module federation, or other similar mechanisms in other bundlers/module loaders, such as rspack and System.js.
```

Comment:
At some point in this document we should link to the upstream docs: https://webpack.js.org/concepts/module-federation/
As a frontend ignoramus I had not realized just by reading this OEP that Webpack module federation is a concept that already exists outside of the Open edX world.

Reply:
Personally I don't like the term "module federation" as an umbrella term for this initiative, because it's webpack-specific and kind of vague. There are various terms but they're all build around the concept of [dynamic import](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) which is now built into the JavaScript language, and when used with a bundler like webpack, Vite, etc. facilitates "lazy loading", "code splitting", and "dependency optimization" in the resulting bundle(s).

I think "runtime module loading" as written two lines above is a much better name. Or even "runtime module loading and shared dependencies" if necessary :p

Reply:
(Hear you loud and clear @regisb - I'll make that clearer)

FWIW, "module federation" first entered our vernacular as a plugin added to webpack, but that [implementation](https://module-federation.io/) has since been pulled out of webpack so that it can be bundler agnostic.  rspack, for instance, uses the same implementation as webpack now, from what I understand.  

I think what we want is composability, and that's what the OEP should be (and is) titled.  I think the way we do that is module federation - which I don't consider to be webpack-specific - which this PR and the OEP mention as an approach, and the implementation is likely via webpack, with respect to our separate Vite conversations. 😄 

I tried to thread the needle between these terms in here, but you both have shown me it could use some clarifying. :)

Reply:
Okay, I rewrote this whole section to try to be a lot clearer about the approach/implementation/etc.  I took @bradenmacdonald's advice and tried to use "runtime module loading" more frequently, except when I'm talking specifically about implementing that via webpack module federation.  Let me know what you both think.

Reply:
Thanks - I think it's definitely more clear now.